src/atomic_kv32.rs

```diff
@@ -0,0 +1,117 @@
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+
+/// An AtomicKV32 is an atomically-updatable pair of two u32s:
+/// a key and a value.
+///
+/// When the key changes, an AtomicKV32 lets the *last* write win,
+/// but, otherwise, the *first* write wins for a series of updates
+/// with the same key (and potentially different values).
+pub(crate) struct AtomicKV32 {
+    // The low half is the key, high half the value.
+    value: AtomicU64,
+}
+
+impl AtomicKV32 {
+    // Creates a fresh AtomicKV32 with this key and value.
+    pub const fn new(key: u32, value: u32) -> AtomicKV32 {
+        AtomicKV32 {
+            value: AtomicU64::new(Self::encode(key, value)),
+        }
+    }
+
+    // Returns the current key-value pair stored in this `AtomicKV32`.
+    //
+    // The first element of the tuple is the key, the second the value.
+    pub fn get(&self) -> (u32, u32) {
+        // Acquire matches the compare_exchange in `set`.
+        Self::decode(self.value.load(Ordering::Acquire))
```

Comment:
this could be relaxed in practice (the value only changes after a fork, and we'll notice stale pids are off), but acquire isn't that expensive.